src/type_tracer/analyzer.rs

```diff
@@ -134,21 +134,142 @@ impl std::fmt::Debug for SymbolId {
   }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[derive(Clone)]
+pub struct NodeRefBox<T> {
+  // the parsed source needs to be kept alive for the duration of the value
+  source: ParsedSource,
+  value: *const T,
+}
+
+impl<T> std::fmt::Debug for NodeRefBox<T> {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    f.debug_struct("DeclRef")
+      .field("value", &"<omitted>")
+      .finish()
+  }
+}
+
+impl<T> NodeRefBox<T> {
+  /// WARNING: Ensure that T is a reference inside ParsedSource. Otherwise
+  /// this is entirely unsafe.
+  fn unsafe_new(parsed_source: &ParsedSource, value: &T) -> Self {
```

Comment:
Maybe this should be an `unsafe` fn and you can ensure `T` inside it?

Reply:
I would for sure if this was public, but since it's private to the module, it's less verbose to just rely on `unsafe` in the name.

What do you mean by "and you can ensure `T` inside it?"?

Reply:
> What do you mean by "and you can ensure `T` inside it?"?

I meant this part `Ensure that T is a reference inside ParsedSource. Otherwise this is entirely unsafe.`. But if that's private I think it's fine.